addons/api/addon/workers/shared-service.js

```diff
@@ -0,0 +1,241 @@
+      const locks = await navigator.locks.query();
+      const isProvider = !locks.held?.find((lock) => lock.name === serviceName);
+      if (!isProvider) {
+        await onNotProvider();
+      }
+      return isProvider;
+    })(),
+  };
+
+  // This lock is used to ensure that only one tab can become the provider with an indefinite lock.
+  navigator.locks.request(serviceName, { mode: 'exclusive' }, async () => {
+    await onBecomeProvider();
+    await new Promise(() => {});
+  });
+
+  const onNotProvider = async () => {
+    const clientId = getClientId();
+    clientChannel = new BroadcastChannel(
+      getClientBroadcastChannelName(clientId, serviceName),
+    );
+
+    const register = async () =>
+      new Promise((resolve) => {
+        const onRegisteredListener = ({ data }) => {
+          if (data.clientId === clientId && data.type === 'registered') {
+            commonChannel.removeEventListener('message', onRegisteredListener);
+            resolve();
+          }
+        };
+
+        // Add an event listener to check for the registration response
+        commonChannel.addEventListener('message', onRegisteredListener);
+        commonChannel.postMessage({ type: 'register', clientId });
+      });
+
+    commonChannel.addEventListener('message', async ({ data }) => {
+      // Check if this was a provider change. If so and we're not the new provider, re-register.
+      if (data.type === 'providerChange' && !(await status.isServiceProvider)) {
+        await onProviderChange?.(await status.isServiceProvider);
+        await register();
+
+        // If there were requests in flight when the provider was changed,
+        // we need to resend them to the new provider.
+        if (requestsInFlight.size > 0) {
+          requestsInFlight.forEach(
+            ({ method, args, resolve, reject }, uuid) => {
+              requestFromProvider(uuid, resolve, reject, method, args);
+            },
+          );
+        }
+      }
+    });
+
+    await register();
+
+    readyResolve?.();
+    readyResolve = null;
+  };
+
+  const onBecomeProvider = async () => {
+    status.isServiceProvider = Promise.resolve(true);
+
+    if (readyResolve === null) {
+      status.ready = new Promise((resolve) => {
+        readyResolve = resolve;
+      });
+    }
+
+    // Set an event listener for other tabs to register against with the provider tab
+    commonChannel.addEventListener('message', async (event) => {
+      const { clientId, type } = event.data;
+      if (type !== 'register') {
+        return;
+      }
+
+      const clientChannel = new BroadcastChannel(
+        getClientBroadcastChannelName(clientId, serviceName),
+      );
+
+      // Acquire another lock that immediately is waiting for the
+      // initial lock on the client tab to be resolved to handle cleanup.
+      navigator.locks.request(clientId, { mode: 'exclusive' }, async () => {
+        // The client tab was closed, clean up.
```

Comment:
We're defining a client tab with an active sqlite connection as "resolved" to mean the tab was closed, but not when the user switches to another tab?